File: speechService.js
const sdk = require('microsoft-cognitiveservices-speech-sdk');

class SpeechService {
    constructor() {
        this.speechConfig = null;
        this.recognizer = null;
        this.synthesizer = null;
        this.isRecognizing = false;
        this.pushStream = null; // for real-time audio input
        this.initializeSpeechConfig();
    }

    initializeSpeechConfig() {
        try {
            if (process.env.NODE_ENV === 'test') {
                console.log('Skipping speech service configuration in test environment');
                this.speechConfig = {
                    speechRecognitionLanguage: 'en-US',
                    speechSynthesisVoiceName: 'en-US-JennyNeural'
                };
                return;
            }

            const speechKey = process.env.SPEECH_SERVICE_KEY;
            const speechRegion = process.env.SPEECH_SERVICE_REGION;

            if (!speechKey || !speechRegion) {
                throw new Error('Speech service key and region must be configured');
            }

            this.speechConfig = sdk.SpeechConfig.fromSubscription(speechKey, speechRegion);
            this.speechConfig.speechRecognitionLanguage = process.env.SPEECH_LANGUAGE || 'en-US';
            this.speechConfig.speechSynthesisVoiceName = process.env.SPEECH_VOICE_NAME || 'en-US-JennyNeural';
            this.speechConfig.speechSynthesisOutputFormat =
                sdk.SpeechSynthesisOutputFormat.Audio16Khz32KBitRateMonoMp3;

            console.log('Speech service configuration initialized', {
                region: speechRegion,
                language: this.speechConfig.speechRecognitionLanguage,
                voice: this.speechConfig.speechSynthesisVoiceName
            });

        } catch (error) {
            console.error('Failed to initialize speech configuration', {
                error: error.message,
                stack: error.stack
            });
            throw error;
        }
    }

    async speechToText(audioBuffer) {
        return new Promise((resolve, reject) => {
            try {
                if (process.env.NODE_ENV === 'test') {
                    setTimeout(() => resolve('Hello world'), 10);
                    return;
                }

                if (!this.speechConfig) {
                    reject(new Error('Speech configuration not initialized'));
                    return;
                }

                const pushStream = sdk.AudioInputStream.createPushStream();
                pushStream.write(audioBuffer);
                pushStream.close();

                const audioConfig = sdk.AudioConfig.fromStreamInput(pushStream);
                const recognizer = new sdk.SpeechRecognizer(this.speechConfig, audioConfig);

                recognizer.recognizeOnceAsync(
                    (result) => {
                        recognizer.close();
                        resolve(result.text);
                    },
                    (error) => {
                        recognizer.close();
                        reject(error);
                    }
                );

            } catch (error) {
                console.error('Error in speechToText', { error: error.message });
                reject(error);
            }
        });
    }

    async startContinuousRecognition(onRecognizedCallback, onErrorCallback = null) {
        try {
            if (this.isRecognizing) {
                console.warn('Continuous recognition is already running');
                return;
            }
            if (!this.speechConfig) {
                throw new Error('Speech configuration not initialized');
            }

            this.pushStream = sdk.AudioInputStream.createPushStream();
            const audioConfig = sdk.AudioConfig.fromStreamInput(this.pushStream);

            this.recognizer = new sdk.SpeechRecognizer(this.speechConfig, audioConfig);

            this.recognizer.recognized = (s, e) => {
                if (e.result.reason === sdk.ResultReason.RecognizedSpeech) {
                    console.log('Continuous speech recognized', { text: e.result.text });
                    if (onRecognizedCallback) onRecognizedCallback(e.result.text);
                }
            };

            this.recognizer.canceled = (s, e) => {
                console.error('Continuous recognition canceled', {
                    reason: e.reason,
                    errorDetails: e.errorDetails
                });
                this.isRecognizing = false;
                if (onErrorCallback) onErrorCallback(new Error(`Recognition canceled: ${e.errorDetails}`));
            };

            this.recognizer.sessionStopped = () => {
                console.log('Continuous recognition session stopped');
                this.isRecognizing = false;
            };

            this.recognizer.startContinuousRecognitionAsync(
                () => {
                    console.log('Continuous speech recognition started');
                    this.isRecognizing = true;
                },
                (error) => {
                    console.error('Failed to start continuous recognition', { error: error.message });
                    if (onErrorCallback) onErrorCallback(error);
                }
            );

        } catch (error) {
            console.error('Error starting continuous recognition', { error: error.message });
            throw error;
        }
    }

    feedAudioChunk(chunk) {
        if (this.pushStream) {
            this.pushStream.write(chunk);
        } else {
            console.warn('No active push stream to feed audio');
        }
    }

    async stopContinuousRecognition() {
        try {
            if (!this.isRecognizing || !this.recognizer) {
                console.warn('No continuous recognition to stop');
                return;
            }

            this.recognizer.stopContinuousRecognitionAsync(
                () => {
                    console.log('Continuous recognition stopped');
                    this.isRecognizing = false;
                    this.recognizer.close();
                    this.recognizer = null;
                    if (this.pushStream) {
                        this.pushStream.close();
                        this.pushStream = null;
                    }
                },
                (error) => {
                    console.error('Error stopping continuous recognition', { error: error.message });
                    this.isRecognizing = false;
                }
            );

        } catch (error) {
            console.error('Error in stopContinuousRecognition', { error: error.message });
            throw error;
        }
    }

    async textToSpeech(text, voiceName = null) {
        return new Promise((resolve, reject) => {
            try {
                if (process.env.NODE_ENV === 'test') {
                    setTimeout(() => resolve(Buffer.from('fake audio data')), 10);
                    return;
                }

                if (!this.speechConfig || !text) {
                    reject(new Error('Speech configuration not initialized or text is empty'));
                    return;
                }

                if (voiceName) {
                    this.speechConfig.speechSynthesisVoiceName = voiceName;
                }

                const synthesizer = new sdk.SpeechSynthesizer(this.speechConfig, null);
                const ssml = this.generateSSML(text, voiceName);

                synthesizer.speakSsmlAsync(
                    (ssml, result) => {
                        if (result.reason === sdk.ResultReason.SynthesizingAudioCompleted) {
                            const audioBuffer = Buffer.from(result.audioData);
                            resolve(audioBuffer);
                        } else {
                            reject(new Error(`Speech synthesis failed: ${result.errorDetails}`));
                        }
                        synthesizer.close();
                    },
                    (error) => {
                        synthesizer.close();
                        reject(error);
                    }
                );
            } catch (error) {
                console.error('Error in textToSpeech', { error: error.message });
                reject(error);
            }
        });
    }

    generateSSML(text, voiceName = null) {
        const voice = voiceName || this.speechConfig.speechSynthesisVoiceName;
        const cleanText = text.replace(/[<>&"']/g, (match) => ({
            '<': '&lt;',
            '>': '&gt;',
            '&': '&amp;',
            '"': '&quot;',
            "'": '&apos;'
        }[match]));
        return `
            <speak version="1.0" xmlns="http://www.w3.org/2001/10/synthesis" xml:lang="en-US">
                <voice name="${voice}">
                    <prosody rate="medium" pitch="medium">${cleanText}</prosody>
                </voice>
            </speak>
        `;
    }

    async saveAudioToFile(audioBuffer, filePath) {
        try {
            const fs = require('fs').promises;
            await fs.writeFile(filePath, audioBuffer);
            console.log('Audio saved to file', { filePath, size: audioBuffer.length });
        } catch (error) {
            console.error('Error saving audio to file', { error: error.message, filePath });
            throw error;
        }
    }

    dispose() {
        try {
            if (this.isRecognizing) {
                this.stopContinuousRecognition();
            }
            if (this.synthesizer) {
                this.synthesizer.close();
                this.synthesizer = null;
            }
            if (this.speechConfig) {
                this.speechConfig = null;
            }
            console.log('Speech service disposed');
        } catch (error) {
            console.error('Error disposing speech service', { error: error.message });
        }
    }
}

module.exports = { SpeechService };

File: Services/SpeechService.cs
using Microsoft.CognitiveServices.Speech;
using Microsoft.CognitiveServices.Speech.Audio;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace VoxRepoAgent.Services
{
    public class SpeechService : ISpeechService, IDisposable
    {
        private readonly ILogger<SpeechService> _logger;
        private readonly IConfiguration _configuration;
        private SpeechConfig? _speechConfig;
        private SpeechRecognizer? _recognizer;
        private SpeechSynthesizer? _synthesizer;
        private bool _isRecognizing;
        private PushAudioInputStream? _pushStream;
        private bool _disposedValue;

        public SpeechService(ILogger<SpeechService> logger, IConfiguration configuration)
        {
            _logger = logger;
            _configuration = configuration;
            InitializeSpeechConfig();
        }

        private void InitializeSpeechConfig()
        {
            try
            {
                var environment = _configuration["ASPNETCORE_ENVIRONMENT"] ?? Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");
                if (environment == "Test")
                {
                    _logger.LogInformation("Skipping speech service configuration in test environment");
                    return;
                }

                var speechKey = _configuration["SPEECH_SERVICE_KEY"] ?? Environment.GetEnvironmentVariable("SPEECH_SERVICE_KEY");
                var speechRegion = _configuration["SPEECH_SERVICE_REGION"] ?? Environment.GetEnvironmentVariable("SPEECH_SERVICE_REGION");

                if (string.IsNullOrEmpty(speechKey) || string.IsNullOrEmpty(speechRegion))
                {
                    throw new InvalidOperationException("Speech service key and region must be configured");
                }

                _speechConfig = SpeechConfig.FromSubscription(speechKey, speechRegion);
                _speechConfig.SpeechRecognitionLanguage = _configuration["SPEECH_LANGUAGE"] ?? Environment.GetEnvironmentVariable("SPEECH_LANGUAGE") ?? "en-US";
                _speechConfig.SpeechSynthesisVoiceName = _configuration["SPEECH_VOICE_NAME"] ?? Environment.GetEnvironmentVariable("SPEECH_VOICE_NAME") ?? "en-US-JennyNeural";
                _speechConfig.SetSpeechSynthesisOutputFormat(SpeechSynthesisOutputFormat.Audio16Khz32KBitRateMonoMp3);

                _logger.LogInformation("Speech service configuration initialized - Region: {Region}, Language: {Language}, Voice: {Voice}",
                    speechRegion, _speechConfig.SpeechRecognitionLanguage, _speechConfig.SpeechSynthesisVoiceName);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to initialize speech configuration");
                throw;
            }
        }

        public async Task<string> SpeechToTextAsync(byte[] audioBuffer)
        {
            try
            {
                var environment = _configuration["ASPNETCORE_ENVIRONMENT"] ?? Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");
                if (environment == "Test")
                {
                    await Task.Delay(10);
                    return "Hello world";
                }

                if (_speechConfig == null)
                {
                    throw new InvalidOperationException("Speech configuration not initialized");
                }

                if (audioBuffer == null || audioBuffer.Length == 0)
                {
                    throw new ArgumentException("Audio buffer must not be null or empty", nameof(audioBuffer));
                }

                using var pushStream = AudioInputStream.CreatePushStream();
                pushStream.Write(audioBuffer);
                pushStream.Close();

                using var audioConfig = AudioConfig.FromStreamInput(pushStream);
                using var recognizer = new SpeechRecognizer(_speechConfig, audioConfig);

                var result = await recognizer.RecognizeOnceAsync();

                if (result.Reason == ResultReason.RecognizedSpeech)
                {
                    return result.Text;
                }
                else if (result.Reason == ResultReason.NoMatch)
                {
                    _logger.LogWarning("No speech could be recognized");
                    return string.Empty;
                }
                else if (result.Reason == ResultReason.Canceled)
                {
                    var cancellation = CancellationDetails.FromResult(result);
                    _logger.LogError("Speech recognition was cancelled: {Reason} - {ErrorDetails}", cancellation.Reason, cancellation.ErrorDetails);
                    throw new InvalidOperationException($"Speech recognition was cancelled: {cancellation.ErrorDetails}");
                }

                return string.Empty;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error in speech to text conversion");
                throw;
            }
        }

        public async Task StartContinuousRecognitionAsync(Func<string, Task> onRecognizedCallback, Func<Exception, Task>? onErrorCallback = null)
        {
            try
            {
                if (_isRecognizing)
                {
                    _logger.LogWarning("Continuous recognition is already running");
                    return;
                }

                if (_speechConfig == null)
                {
                    throw new InvalidOperationException("Speech configuration not initialized");
                }

                _pushStream = AudioInputStream.CreatePushStream();
                var audioConfig = AudioConfig.FromStreamInput(_pushStream);
                _recognizer = new SpeechRecognizer(_speechConfig, audioConfig);

                _recognizer.Recognizing += (sender, e) =>
                {
                    if (!string.IsNullOrEmpty(e.Result.Text))
                    {
                        _logger.LogDebug("Interim recognizing: {Text}", e.Result.Text);
                    }
                };

                _recognizer.Recognized += async (sender, e) =>
                {
                    if (e.Result.Reason == ResultReason.RecognizedSpeech && !string.IsNullOrEmpty(e.Result.Text))
                    {
                        _logger.LogInformation("Continuous speech recognized: {Text}", e.Result.Text);
                        await onRecognizedCallback(e.Result.Text);
                    }
                    else if (e.Result.Reason == ResultReason.NoMatch)
                    {
                        _logger.LogWarning("No match for audio chunk");
                    }
                };

                _recognizer.Canceled += async (sender, e) =>
                {
                    _logger.LogError("Continuous recognition canceled: {Reason} - {ErrorDetails}", e.Reason, e.ErrorDetails);
                    _isRecognizing = false;

                    DisposeRecognizer();

                    if (onErrorCallback != null)
                    {
                        await onErrorCallback(new InvalidOperationException($"Recognition canceled: {e.ErrorDetails ?? e.Reason.ToString()}"));
                    }
                };

                _recognizer.SessionStopped += (sender, e) =>
                {
                    _logger.LogInformation("Continuous recognition session stopped");
                    _isRecognizing = false;
                    DisposeStreamResources();
                };

                await _recognizer.StartContinuousRecognitionAsync();
                _logger.LogInformation("Continuous speech recognition started");
                _isRecognizing = true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error starting continuous recognition");
                throw;
            }
        }

        public async Task FeedAudioChunkAsync(byte[] chunk)
        {
            if (_pushStream == null)
            {
                _logger.LogWarning("No active push stream to feed audio");
                return;
            }

            try
            {
                if (chunk == null || chunk.Length == 0)
                {
                    _logger.LogWarning("Empty chunk passed to FeedAudioChunkAsync");
                    return;
                }

                _pushStream.Write(chunk);
                await Task.CompletedTask;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error writing audio chunk to push stream");
            }
        }

        public async Task StopContinuousRecognitionAsync()
        {
            try
            {
                if (!_isRecognizing || _recognizer == null)
                {
                    _logger.LogWarning("No continuous recognition to stop");
                    DisposeStreamResources();
                    return;
                }

                await _recognizer.StopContinuousRecognitionAsync();
                _logger.LogInformation("Continuous recognition stopped");
                _isRecognizing = false;
                DisposeRecognizer();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error stopping continuous recognition");
                _isRecognizing = false;
                throw;
            }
        }

        public async Task<byte[]> TextToSpeechAsync(string text, string? voiceName = null)
        {
            try
            {
                var environment = _configuration["ASPNETCORE_ENVIRONMENT"] ?? Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");
                if (environment == "Test")
                {
                    await Task.Delay(10);
                    return "fake audio data"u8.ToArray();
                }

                if (_speechConfig == null || string.IsNullOrEmpty(text))
                {
                    throw new InvalidOperationException("Speech configuration not initialized or text is empty");
                }

                if (!string.IsNullOrEmpty(voiceName))
                {
                    _speechConfig.SpeechSynthesisVoiceName = voiceName;
                }

                using var synthesizer = new SpeechSynthesizer(_speechConfig, null);
                var ssml = GenerateSSML(text, voiceName);

                var result = await synthesizer.SpeakSsmlAsync(ssml);

                if (result.Reason == ResultReason.SynthesizingAudioCompleted)
                {
                    return result.AudioData;
                }
                else if (result.Reason == ResultReason.Canceled)
                {
                    var cancellation = SpeechSynthesisCancellationDetails.FromResult(result);
                    throw new InvalidOperationException($"Speech synthesis failed: {cancellation.ErrorDetails}");
                }

                throw new InvalidOperationException($"Speech synthesis failed with reason: {result.Reason}");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error in text to speech conversion");
                throw;
            }
        }

        private string GenerateSSML(string text, string? voiceName = null)
        {
            var voice = voiceName ?? _speechConfig?.SpeechSynthesisVoiceName ?? "en-US-JennyNeural";
            var cleanText = System.Net.WebUtility.HtmlEncode(text);
            
            return $@"
                <speak version=""1.0"" xmlns=""http://www.w3.org/2001/10/synthesis"" xml:lang=""en-US"">
                    <voice name=""{voice}"">
                        <prosody rate=""medium"" pitch=""medium"">{cleanText}</prosody>
                    </voice>
                </speak>";
        }

        public async Task SaveAudioToFileAsync(byte[] audioBuffer, string filePath)
        {
            try
            {
                await File.WriteAllBytesAsync(filePath, audioBuffer);
                _logger.LogInformation("Audio saved to file: {FilePath} (size: {Size} bytes)", filePath, audioBuffer.Length);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error saving audio to file: {FilePath}", filePath);
                throw;
            }
        }

        private void DisposeRecognizer()
        {
            try
            {
                _recognizer?.Dispose();
                _recognizer = null;
                DisposeStreamResources();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error disposing recognizer");
            }
        }

        private void DisposeStreamResources()
        {
            try
            {
                _pushStream?.Close();
                _pushStream = null;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error disposing stream resources");
            }
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!_disposedValue)
            {
                if (disposing)
                {
                    try
                    {
                        if (_isRecognizing)
                        {
                            StopContinuousRecognitionAsync().Wait(TimeSpan.FromSeconds(5));
                        }
                        
                        _synthesizer?.Dispose();
                        _synthesizer = null;
                        _speechConfig = null;
                        
                        _logger.LogInformation("Speech service disposed");
                    }
                    catch (Exception ex)
                    {
                        _logger.LogError(ex, "Error disposing speech service");
                    }
                }

                _disposedValue = true;
            }
        }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }
    }
}